src/riak_kv_qry_compiler.erl

```diff
@@ -56,45 +56,129 @@ comp2(#ddl_v1{} = DDL, #riak_sql_v1{is_executable = false,
+
+%%
+compile_select(DDL, SelectSpec1) ->
+    % compile each select column and put all the results types into a set, if
+    % any of the results are aggregate then aggregate is the result type for the
+    % whole query
+    CompileColFn = 
+        fun(ColX, {TypeSetX, InitColStatesX, SelectSpecX}) ->
+            {ColResultType, ColStateX, X} = compile_select_col(DDL, ColX),
+            {sets:add_element(ColResultType, TypeSetX), % result types
+             [ColStateX | InitColStatesX], % initial states if aggregates
+             [X | SelectSpecX]} % compiled select fun
+        end,
+    {TypeSet, InitialState, SelectSpec2} =
+        lists:foldr(CompileColFn, {sets:new(), [], []}, SelectSpec1),
+    case sets:is_element(aggregate, TypeSet) of
+        true  -> {aggregate, InitialState, SelectSpec2};
+        false -> {rows, [], SelectSpec2}
+    end.
 
 %% Compile a single selection column into a fun that can extract the cell 
 %% from the row.
--spec compile_select(DDL::#ddl_v1{}, ColumnName::[binary()]) ->
+-spec compile_select_col(DDL::#ddl_v1{}, ColumnSpec::{identifier,[binary()]}) ->
         compiled_select().
-compile_select(_, [<<"*">>]) ->
-    fun(Row) ->
-        Row
+compile_select_col(DDL, {funcall, {FnName, [Arg1]}} = Select) ->
+    Initial_state = riak_ql_window_agg_fns:start_state(FnName),
+    Compiled_arg1 = compile_select_col_stateless(DDL, Arg1),
+    case riak_ql_window_agg_fns:start_state(FnName) of
+        stateless ->
+            {rows, undefined, compile_select_col_stateless(DDL, Select)};
+        Initial_state ->
+            Fn =
+                fun(Row, State) ->
+                    riak_ql_window_agg_fns:FnName(Compiled_arg1(Row), State)
+                end,
+            {aggregate, Initial_state, Fn}
     end;
-compile_select(#ddl_v1{ fields = Fields }, ColumnName) ->
-    Index = field_index_of(Fields, ColumnName),
+compile_select_col(DDL, Select) ->
+    Fn = compile_select_col_stateless(DDL, Select),
+    % to support aggregates that have a running state all top level funs must be
+    % arity two, when we know the function is stateless wrap it in a two arity
+    % fun and ignore the state arg
+    {rows, undefined, fun(Row, _) -> Fn(Row) end}.
+
+%% Returns a one arity fun which is stateless for example pulling a field from a
+%% row.
+compile_select_col_stateless(_, {identifier, [<<"*">>]}) ->
+    fun(Row) -> Row end;
+compile_select_col_stateless(_, {Type, V})
+        when Type == integer; Type == varchar; Type == double; Type == boolean ->
+    fun(_) -> V end;
+compile_select_col_stateless(#ddl_v1{ fields = Fields }, {identifier, ColumnName}) ->
+    Index = column_index_of(Fields, to_column_name_binary(ColumnName)),
     fun(Row) ->
         lists:nth(Index, Row)
-    end.
+    end;
+compile_select_col_stateless(DDL, {Op, A, B}) ->
+    Arg_a = compile_select_col_stateless(DDL, A),
+    Arg_b = compile_select_col_stateless(DDL, B),
+    compile_select_col_stateless2(Op, Arg_a, Arg_b).
+
```

Comment:
Can we also get a function body here to do `negate` c.f. https://bashoeng.atlassian.net/browse/RTS-753 ?

```erlang
compile_select_col_stateless(DDL, {negate, A}) ->
  ToBeNegated = compile_select_col_stateless(DDL, A),
  fun(Row) -> -ToBeNegated(Row) end.
```